home/utils.py

```diff
@@ -67,6 +70,26 @@
 ]
 
 
+def cosine_similarity(A, B):
+    return float((np.dot(A, B) / (np.linalg.norm(A) * np.linalg.norm(B))))
+
+
+def retrieve_top_n_content_pieces(user_input, queryset, n=5, content_type=None):
+    # similar_embeddings = [{'faq_name':, 'faq_content':, 'embedding':}, ...] # We need to filter by content type and then retrieve their embeddings
+    # Generate embedding for user text
+    user_embedding = model.encode([user_input])
```

Comment:
Can we update user_embedding to include text preprocessing:

user_embedding = model.encode([emoji_pattern.sub(r"", user_input).replace("\n\n", " ").replace("*", "")])

Reply:
@PaulEloffPraekelt Is this comment still relevant? I remember you saying something about the newer models not needing preprocessing?

Reply:
@KaitCrawford let's please exclude the preprocessing that I initially suggested to add above. Can we please have a catchup sometime to align on where you are at? Apart from the confidence threshold mentioned in my other comment, there may also be another line or two of our code that I would like to change before implementation. Also, I would like to ensure that, whichever way we are storing the data, we will be able to use it as intended downstream?

Reply:
@PaulEloffPraekelt I'm currently looking to add profiling to the endpoint/codebase so that we can monitor how having the model here affects the rest of the app.
But I need to deploy this sometime this sprint as well. So let's have that conversation sooner rather than later please

Reply:
@PaulEloffPraekelt we added text preprocessing to the pages when we calculate and save their embeddings but I see now we didn't add it here on the user input. Can you confirm that we still need it here as well? (eg dealing with emojis in user questions)